creator/studies/data_generator/study_generator.py

```diff
@@ -0,0 +1,636 @@
+        of the ingest pipeline, called in run_ingest_pipeline
+        """
+        dry_run = ingest_kwargs.get("dry_run")
+
+        if not random_seed:
+            ra.seed(0)
+
+        # Cleanup all previously written data and data loaded into Dataservice
+        if clean:
+            self.clean(dry_run=dry_run)
+
+        # Create study data files
+        self.generate_files()
+
+        # Ingest study, seq center, and study data files into Dataservice
+        self.run_ingest_pipeline(**ingest_kwargs)
+
+        # Ensure all kf_ids were loaded and entity counts are valid
+        if verify_counts:
+            self._verify_counts(dry_run=dry_run)
+
+    def clean(self, dry_run: bool = False, all_studies: bool = False) -> None:
+        """
+        Clean up data before ingestion begins
+
+        1. Delete study data in Dataservice (if not a dry run ingest)
+        2. Delete existing ingest package with data files
+
+        :param all_studies: Whether to delete all studies' data in
+        Dataservice or just the entities linked to StudyGenerator.study_id
+        :param dry_run: Whether its a dry run ingest, and therefore does not
+        require deleting study data from dataservice
+        """
+        logger.info("Clean up before study generation begins")
+
+        # Delete entities in Dataservice
+        if not dry_run:
+            study_ids = None if all_studies else [self.study_id]
+            delete_entities(self.dataservice_url, study_ids=study_ids)
+
+        # Delete existing data files
+        logger.info(f"Deleting files in {self.ingest_package_dir}")
+        shutil.rmtree(self.ingest_package_dir, ignore_errors=True)
+
+    def generate_files(self) -> None:
+        """
+        Generate clinical and genomic data files for a test study.
+        Read files into DataFrames if they exist, otherwise create them:
+
+        1. bio_manifest.tsv
+        2. sequencing_manifest.tsv
+        3. s3_source_gf_manifest.tsv
+        4. s3_harmonized_gf_manifest.tsv
+        5. gwo_manifest.tsv
+
+        Files are written to:
+        <StudyGenerator.working_dir>/<StudyGenerator.study_id>_ingest_package/data # noqa
+        """
+        self.expected_counts = {
+            "families": 2,
+            "participants": self.total_specimens,
+            "biospecimens": self.total_specimens,
+            "sequencing-experiments": self.total_specimens,
+            # Include only source gfs since harmonized gfs don't get loaded
+            # by study generator
+            "genomic-files": self.total_specimens * len(SOURCE_GF_EXTS),
+        }
+        logger.info(
+            f"Creating clinical and genomic files for {self.study_id}. "
+            f"Entities:\n{pformat(self.expected_counts)}"
+        )
+
+        # Try read from disk
+        fps = [os.path.join(self.data_dir, fn) for fn in self._df_creators]
+        read_existing = all([os.path.exists(fp) for fp in fps])
```

Comment:
```suggestion
fps = (os.path.join(self.data_dir, fn) for fn in self._df_creators)
read_existing = all((os.path.exists(fp) for fp in fps))
```

Reply:
Lol the iterable is so tiny that using a generator exp prob doesn't matter?

Reply:
I knowwww it's just a habit